File: wrc_erpnext/wrc_erpnext/validations/work_order.py
# Copyright (c) 2015, Frappe Technologies Pvt. Ltd. and Contributors
# License: GNU General Public License v3. See license.txt

from __future__ import unicode_literals
import frappe
from frappe import _
from frappe.utils import nowdate
from erpnext.manufacturing.doctype.work_order.work_order import check_if_scrap_warehouse_mandatory

@frappe.whitelist()
def get_bom_barcode_details(barcode, project=None):
	bom = frappe.db.get_value("BOM Barcode", {"barcode": barcode}, "parent")

	if not bom:
		frappe.msgprint(
			_("Invalid Barcode. There is no BOM attached to this barcode."))
		return "invalid"

	else:
		item = frappe.db.get_value("BOM", bom, "item")
		res = set_bom_details(item)

		if project:
			if not frappe.db.exists("BOM", {"name": bom, "project": project}):
				res = get_bom_barcode_details(barcode)
				frappe.msgprint(_("Default BOM not found for Item {0} and Project {1}").format(
					item, project), alert=1)

		bom_data = frappe.db.get_value('BOM', bom,
                                 ['project', 'allow_alternative_item', 'transfer_material_against', 'item_name'], as_dict=1)
		res['project'] = project or bom_data.pop("project")
		res.update(bom_data)
		res['item'] = item
		res['bom_no'] = bom
		res.update(check_if_scrap_warehouse_mandatory(res["bom_no"]))
		return res

@frappe.whitelist()
def get_item_details(item, project=None):
	res = set_bom_details(item)

	filters = {"item": item, "is_default": 1}

	if project:
		filters = {"item": item, "project": project}

	res["bom_no"] = frappe.db.get_value("BOM", filters=filters)

	if not res["bom_no"]:
		variant_of = frappe.db.get_value("Item", item, "variant_of")

		if variant_of:
			res["bom_no"] = frappe.db.get_value(
				"BOM", filters={"item": variant_of, "is_default": 1})

	if not res["bom_no"]:
		if project:
			res = wrc_erpnext.get_item_details(item)
			frappe.msgprint(_("Default BOM not found for Item {0} and Project {1}").format(
				item, project), alert=1)
		else:
			frappe.throw(_("Default BOM for {0} not found").format(item))

	bom_data = frappe.db.get_value('BOM', res['bom_no'],
								['project', 'allow_alternative_item', 'transfer_material_against', 'item_name'], as_dict=1)

	barcode = frappe.db.get_value(
		"BOM Barcode", {"parent": res['bom_no']}, "barcode")
	res['project'] = project or bom_data.pop("project")
	res['bom_barcode'] = barcode
	res.update(bom_data)
	res.update(check_if_scrap_warehouse_mandatory(res["bom_no"]))

	return res

def set_bom_details(item):
	res = frappe.db.sql("""
		select stock_uom, description
		from `tabItem`
		where disabled=0
			and (end_of_life is null or end_of_life='0000-00-00' or end_of_life > %s)
			and name=%s
	""", (nowdate(), item), as_dict=1)

	if not res:
		return {}
	res = res[0]
	return res

File: wrc_erpnext/public/js/Work Order.js
// Copyright (c) 2017, Frappe Technologies Pvt. Ltd. and contributors
// For license information, please see license.txt

frappe.ui.form.on("Work Order", {
	production_item: function (frm) {
		if (frm.doc.production_item) {
			frappe.call({
				method: "wrc_erpnext.wrc_erpnext.validations.work_order.get_item_details",
				args: {
					item: frm.doc.production_item,
					project: frm.doc.project
				},
				freeze: true,
				callback: function (r) {
					frm.set_value('sales_order', "");
					frm.trigger('set_sales_order');
					erpnext.in_production_item_onchange = true;
					let fields = ["description", "stock_uom", "project", "bom_no", "allow_alternative_item",
						"transfer_material_against", "item_name"];

					$.each(fields, function (i, field) {
						frm.set_value(field, r.message[field]);
						frm.doc.bom_barcode = r.message["bom_barcode"];
					});
					if (r.message["set_scrap_wh_mandatory"]) {
						frm.toggle_reqd("scrap_warehouse", true);
					}
					erpnext.in_production_item_onchange = false;
				}
			});
		}
	},

	bom_barcode: function (frm) {
		frappe.call({
			method: "wrc_erpnext.wrc_erpnext.validations.work_order.get_bom_barcode_details",
			args: {
				barcode: frm.doc.bom_barcode,
				project: frm.doc.project
			},

			callback: function (r) {
				if (r.message !== "invalid") {
					frm.doc.production_item = r.message["item"];
					frm.set_value("item_name", r.message["item_name"]);
					frm.set_value("bom_no", r.message["bom_no"]);

					frm.set_value('sales_order', "");
					frm.trigger('set_sales_order');

					erpnext.in_production_item_onchange = true;
					let fields = ["description", "stock_uom", "project", "allow_alternative_item",
						"transfer_material_against", "item_name"];

					$.each(fields, function (i, field) {
						frm.set_value(field, r.message[field]);
					});

					if (r.message["set_scrap_wh_mandatory"]) {
						frm.toggle_reqd("scrap_warehouse", true);
					}
					frm.refresh();
				}
				else {
					frm.set_value("production_item", "");
					frm.set_value("bom_no", "");
					frm.set_value("quantity", "");
					frm.refresh();
				}
			}
		});
	},
});